### PR TITLE
[Gekidou MM-39733] Port Create Channel from V1

### PR DESCRIPTION
**Pull request by larkox**
#### Summary
Follow up on https://github.com/mattermost/mattermost-mobile/pull/5867
#### Ticket Link
https://mattermost.atlassian.net/browse/MM-39733
#### Screenshots
Screenshot from 2022-03-21 16-01-00
Screenshot from 2022-03-21 16-01-44

#### Release Note
```release-note
None
```

### CODE REVIEW

File: app/actions/remote/channel.ts
Comment (larkox) [inline app/actions/remote/channel.ts:176]: Type and DisplayName are not in this table, so they should be handled.

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
Hey @larkox, I'm not sure what was agreed upon, but can we make some tweaks based on the design done by Marina? 

- Remove the grey background - make it all `center channel bg`
- add a 20px margin so the inputs aren't right to the edge
- Can we make sure all text is using `Open Sans`? It looks like the smaller text elements are not using Open Sans
- Can we make 'Make Private' semi bold and add the lock icon? Also, can we confirm that it's Body 200 (16 font-size) for this title?
- Need to add the 'x' icon to cancel out of close the modal
- The Create button should be `Open Sans` - see the Edit post modal for the an example of how this has been updated

**Sample design:**

**Comment by larkox** [comment]
@matthewbirtch I was mainly focusing on getting the PR back to shape "technically", but all your changes seem reasonable. Will make sure they enter in this PR.

**Comment by larkox** [comment]
@matthewbirtch Beeing looking at this, and I have almost ready most of your requested changes. The only thing that may be better for a different PR is the ones changing the style of the header.

The header is not controlled by the components, but it is more general (the same for all components). Let me know if you see anything different between this and other headers (I should be able to fix that) but changes in general would be something to deal in a different PR.

**Comment by matthewbirtch** [comment]
> The header is not controlled by the components, but it is more general (the same for all components). Let me know if you see anything different between this and other headers (I should be able to fix that) but changes in general would be something to deal in a different PR.

Yep, understood. I think @enahum already updated the header component in the 'Edit post modal' PR, so that's why I was surprised to see that it was not right in this PR.

File: app/components/autocomplete/emoji_suggestion/emoji_suggestion.tsx
Comment (enahum) [inline app/components/autocomplete/emoji_suggestion/emoji_suggestion.tsx:78]: the component does not seem to be receiving this prop from anywhere, do we really need it?

File: app/components/channel_list/header/plus_menu/index.tsx
Comment (enahum) [inline app/components/channel_list/header/plus_menu/index.tsx:40]: instead of passing `closeButton` as a prop you have two options:
1. (preferred) create the left button here and set it as the modal options
2. get the close button in the component were you are setting the modal buttons.

Comment (enahum) [inline app/components/channel_list/header/plus_menu/index.tsx:40]: instead of passing `closeButton` as a prop you have two options:
1. (preferred) create the left button here and set it as the modal options. (there is a helper function to build the button)
2. get the close button in the component were you are setting the modal buttons.

File: app/screens/create_or_edit_channel/create_or_edit_channel.tsx
Comment (enahum) [inline app/screens/create_or_edit_channel/create_or_edit_channel.tsx:78]: Use the `buildNavigationButton` function as it sets the font used in this types of buttons.. in fact you could even extend it to use the text as it seems that I forgot to add it.

Comment (enahum) [inline app/screens/create_or_edit_channel/create_or_edit_channel.tsx:78]: Use the `buildNavigationButton` function as it sets the font used in this types of buttons.. in fact you could even extend it to use the text as it seems that I forgot to add it.

File: app/screens/index.tsx
Comment (enahum) [inline app/screens/index.tsx:139]: nit: can we try and this sorted?

Comment (enahum) [inline app/screens/index.tsx:139]: nit: can we try and this sorted?

File: app/components/floating_text_input_label/index.tsx
Comment (enahum) [inline app/components/floating_text_input_label/index.tsx:163]: no no no no no.. this component is used in the login flow screens that do not really use the theme from the context.

Comment (enahum) [inline app/components/floating_text_input_label/index.tsx:163]: no no no no no.. this component is used in the login flow screens that do not really use the theme from the context.

File: app/screens/edit_profile/components/field.tsx
Comment (enahum) [inline app/screens/edit_profile/components/field.tsx:92]: revert

Comment (enahum) [inline app/screens/edit_profile/components/field.tsx:92]: revert

File: app/screens/edit_server/form.tsx
Comment (enahum) [inline app/screens/edit_server/form.tsx:168]: revert

Comment (enahum) [inline app/screens/edit_server/form.tsx:168]: revert

File: app/screens/login/form.tsx
Comment (enahum) [inline app/screens/login/form.tsx:353]: revert

Comment (enahum) [inline app/screens/login/form.tsx:353]: revert

File: app/screens/mfa/index.tsx
Comment (enahum) [inline app/screens/mfa/index.tsx:238]: revert

Comment (enahum) [inline app/screens/mfa/index.tsx:238]: revert

**Review by enahum — CHANGES_REQUESTED** [review]
the changes in the float input component makes me nervous, we need to test that does changes have no effect on other screens were the component is being used and that is working as expected.

File: app/components/autocomplete/emoji_suggestion/emoji_suggestion.tsx
Comment (larkox) [reply app/components/autocomplete/emoji_suggestion/emoji_suggestion.tsx:78]: We need it to make sure we don't react to the latest post if we are not in the post draft. But it was not wired to the Autocomplete component correctly. Will fix.

**Comment by larkox** [comment]
@enahum Regarding changes to Floating Text breaking other places, I discovered some issues in the original PR in the add server screen that are already addressed. Would be nice to do further testing, but just wanted to let you know that I already looked there.

**Comment by enahum** [comment]
@larkox conflicts after merging #5984

**Review by matthewbirtch — APPROVED** [review]
Looks great @larkox. I'm going to approve this, but if you could fix this one last minor detail as well, that should cover it:

Can we add at least 4 margin between the icon and the `Make private` label?
IMG_8983

File: app/screens/create_or_edit_channel/channel_info_form.tsx
Comment (avinashlng1080) [inline app/screens/create_or_edit_channel/channel_info_form.tsx:167]:
```suggestion
     const blur = useCallback(() => {
        nameInput?.current?.blur();
        purposeInput?.current?.blur();
        headerInput?.current?.blur();
        scrollViewRef.current?.scrollToPosition(0, 0, true);
    }, []);
```

File: app/screens/create_or_edit_channel/channel_info_form.tsx
Comment (avinashlng1080) [inline app/screens/create_or_edit_channel/channel_info_form.tsx:175]:
```suggestion
            scrollViewRef.current?.scrollToPosition(0, headerPosition);
```

File: app/screens/create_or_edit_channel/create_or_edit_channel.tsx
Comment (avinashlng1080) [inline app/screens/create_or_edit_channel/create_or_edit_channel.tsx:127]: You can use the other equivalent form: 

`        const closeButton = CompassIcon.getImageSourceSync('close', 24, theme.sidebarHeaderTextColor);
`

File: app/components/section_item/index.tsx
Comment (avinashlng1080) [inline app/components/section_item/index.tsx:68]:
```suggestion

```

File: app/components/section_item/index.tsx
Comment (avinashlng1080) [inline app/components/section_item/index.tsx:149]: Prefer callback over in-line function call

**Review by avinashlng1080 — CHANGES_REQUESTED** [review]
LGTM.  Some minor nits

File: app/components/section_item/index.tsx
Comment (avinashlng1080) [inline app/components/section_item/index.tsx:67]: Can `testID` be removed if not used ?  If not, then can we assign a default value to the `testID` for if we leave it as `undefined`, then `undefined` will be concatenated to each string in the code below?

File: app/components/section_item/index.tsx
Comment (avinashlng1080) [inline app/components/section_item/index.tsx:116]: shouldn't  theme be part of the dependencies as well ?

File: app/screens/create_or_edit_channel/create_or_edit_channel.tsx
Comment (avinashlng1080) [inline app/screens/create_or_edit_channel/create_or_edit_channel.tsx:8]:
```suggestion
import {Keyboard} from 'react-native';
```

File: app/screens/create_or_edit_channel/create_or_edit_channel.tsx
Comment (avinashlng1080) [inline app/screens/create_or_edit_channel/create_or_edit_channel.tsx:16]:
```suggestion
import ChannelInfoForm from './channel_info_form';
```

and then you would need to sort the imports.

File: app/screens/create_or_edit_channel/create_or_edit_channel.tsx
Comment (avinashlng1080) [inline app/screens/create_or_edit_channel/create_or_edit_channel.tsx:118]:
```suggestion
    }, [editing, theme.sidebarHeaderTextColor, intl, canSave]);
```
..since it would also depends on `formatMessage`

File: app/screens/create_or_edit_channel/channel_info_form.tsx
Comment (avinashlng1080) [inline app/screens/create_or_edit_channel/channel_info_form.tsx:219]: Since the Autocomplete component is not displayed conditionally, you can call it directly in the render method.

File: app/screens/create_or_edit_channel/channel_info_form.tsx
Comment (avinashlng1080) [inline app/screens/create_or_edit_channel/channel_info_form.tsx:340]: Do you need the <View> component?

File: app/screens/create_or_edit_channel/channel_info_form.tsx
Comment (avinashlng1080) [inline app/screens/create_or_edit_channel/channel_info_form.tsx:338]:
```suggestion
```

**Review by avinashlng1080 — CHANGES_REQUESTED** [review]
Looking good; some minor nits.

File: app/screens/create_or_edit_channel/create_or_edit_channel.tsx
Comment (larkox) [reply app/screens/create_or_edit_channel/create_or_edit_channel.tsx:118]: In theory, formatMessage would only change if we are changing the locale. Nevertheless, yes, let's play it safe and use `intl`.

File: app/screens/create_or_edit_channel/channel_info_form.tsx
Comment (larkox) [reply app/screens/create_or_edit_channel/channel_info_form.tsx:340]: Which one? The ScrollView? Why do you think we may not need it?

File: app/actions/remote/channel.ts
Comment (enahum) [inline app/actions/remote/channel.ts:123]: if `DatabaseManager.serverDatabases[serverUrl];` is undefined, the deconstruct here will throw an error

```suggestion
    const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
    if (!operator) {
        return {error: `${serverUrl} database not found`};
    }
```

File: app/actions/remote/channel.ts
Comment (enahum) [inline app/actions/remote/channel.ts:133]:
```suggestion
        const {database} = operator;
        const {currentUserId, currentTeamId} = await getCommonSystemValues(database);
```

File: app/components/channel_list/header/plus_menu/index.tsx
Comment (enahum) [inline app/components/channel_list/header/plus_menu/index.tsx:10]: seems a bit weird.. 0/5 if you want to move the function to another file

File: app/constants/channel.ts
Comment (enahum) [inline app/constants/channel.ts:4]: there is a PR to change this to 1 character in v1 #6093 probably we want that here too

File: app/constants/channel.ts
Comment (enahum) [inline app/constants/channel.ts:9]: Maybe change this to `MIN_CHANNEL_NAME_LENGTH` and the same for max?

File: app/queries/servers/channel.ts
Comment (enahum) [inline app/queries/servers/channel.ts:236]:
```suggestion
export const getCurrentChannelInfo = async (database: Database) => {
```

File: app/screens/create_or_edit_channel/create_or_edit_channel.tsx
Comment (enahum) [inline app/screens/create_or_edit_channel/create_or_edit_channel.tsx:57]: icon is of type `ImageResource`

File: app/screens/create_or_edit_channel/channel_info_form.tsx
Comment (enahum) [reply app/screens/create_or_edit_channel/channel_info_form.tsx:328]: 🤷‍♂️  this looks fine to me

File: app/utils/channel/index.ts
Comment (enahum) [reply app/utils/channel/index.ts:67]: hmm, for me this are all messages, if you want to be picky perhaps call it `errorMessages`

**Review by enahum — CHANGES_REQUESTED** [review]

File: app/screens/create_or_edit_channel/channel_info_form.tsx
Comment (avinashlng1080) [reply app/screens/create_or_edit_channel/channel_info_form.tsx:328]: I mean the wrapping `<View>` around the `<Autocomplete>`

Comment (enahum) [reply app/screens/create_or_edit_channel/channel_info_form.tsx:328]: Because Autocomplete is absolute this view makes it relative to it

File: app/components/channel_list/header/plus_menu/index.tsx
Comment (larkox) [reply app/components/channel_list/header/plus_menu/index.tsx:10]: Now that I think more about it, on the first render we will be setting the button either way, so I will not set the button during creation. That should remove the current weirdness.

File: app/screens/create_or_edit_channel/channel_info_form.tsx
Comment (avinashlng1080) [inline app/screens/create_or_edit_channel/channel_info_form.tsx:89]: to be removed ?

File: app/screens/create_or_edit_channel/channel_info_form.tsx
Comment (avinashlng1080) [inline app/screens/create_or_edit_channel/channel_info_form.tsx:224]: assign a text to it like for others.

File: app/constants/channel.ts
Comment (larkox) [reply app/constants/channel.ts:4]: Made a comment on the original PR that is also relevant here: Should we worry about the server version, or let the server fail?

**Review by avinashlng1080 — COMMENTED** [review]
LGTM and I'm about to approve it.  However, I tested it on an iPhone 13 with iOS 15.4 and the autocomplete component does not appear in the simulator.

**Comment by larkox** [comment]
@avinashlng1080 Just to make sure, the autocomplete should only show on the Header. Name and purpose should not show the autocomplete. Is the autocomplete not showing for you in the header field?

**Comment by avinashlng1080** [comment]
> @avinashlng1080 Just to make sure, the autocomplete should only show on the Header. Name and purpose should not show the autocomplete. Is the autocomplete not showing for you in the header field?

Yes I'm testing the 'header' field.

Simulator Screen Shot - iPhone 13 - 2022-03-29 at 15 00 44

File: app/constants/channel.ts
Comment (enahum) [reply app/constants/channel.ts:4]: That is in fact a very very good question.. I would say if we can identify the error from the server and add a specific error string on the client for that I would let it fail, yes.

Comment (larkox) [reply app/constants/channel.ts:4]: The error string from the server seems good enough, so yes, we can let it fail.
Screenshot from 2022-03-29 13-54-21

**Comment by enahum** [comment]
Conflicts

**Comment by larkox** [comment]
@avinashlng1080 I don't see any obvious reason why it won't work on iPhone. It is working fine on Android emulator. I still cannot develop on iOS, so it would be great if you could take a look.

**Review by enahum — APPROVED** [review]

**Review by avinashlng1080 — APPROVED** [review]

**Pull request merged**
